### ISSUE TITLE
Reduce time it takes to run the examples

### ISSUE THREAD

**Issue by moorepants** [csu-hmc/opty#204]
Here is the current timing from a clean build:

```
computation time summary:
    - ../examples-gallery/plot_pendulum_swing_up_variable_duration.py:   110.06 sec   0.0 MB
    - ../examples-gallery/plot_drone.py:                                  75.58 sec   0.0 MB
    - ../examples-gallery/plot_two_link_pendulum_on_a_cart.py:            75.11 sec   0.0 MB
    - ../examples-gallery/plot_sliding_block.py:                          59.82 sec   0.0 MB
    - ../examples-gallery/plot_parallel_park.py:                          58.78 sec   0.0 MB
    - ../examples-gallery/plot_pendulum_swing_up_fixed_duration.py:       56.85 sec   0.0 MB
    - ../examples-gallery/plot_one_legged_time_trial.py:                  44.69 sec   0.0 MB
    - ../examples-gallery/plot_vyasarayani.py:                            11.22 sec   0.0 MB
    - ../examples-gallery/plot_betts2003.py:                              10.89 sec   0.0 MB
```

Reducing the time it takes to create animations is an easy adjusted, for example. We can safely skip generating a frame for every time instance by skipping frames with a minor adjustment to the code. Here is an example:

https://github.com/csu-hmc/opty/blob/master/examples-gallery/plot_drone.py#L299-L300

**Comment by moorepants** [comment]
Another way to reduce time is to give better initial guesses. We can even store a solution and give that as a guess.

**Comment by Peter230655** [comment]
> Another way to reduce time is to give better initial guesses. We can even store a solution and give that as a guess.

How can I store a solution, so it does not get lost when the program is closed?
I thought about this often, but never came up with an idea... :-(

**Comment by moorepants** [comment]
You can save any value to a txt file with many different methods. NumPy has builtin ways to do this: https://numpy.org/doc/stable/reference/routines.io.html

**Comment by Peter230655** [comment]
> You can save any value to a txt file with many different methods. NumPy has builtin ways to do this: https://numpy.org/doc/stable/reference/routines.io.html

Amazing how this works - and it is two lines of code initially, no additional line in the final simulation. (After your hint, chatGPT told me the lines) I once modeled a drone with propellers, the one that wiggled and it cut down the simulation tine from over one minute to 16 sec. - still wiggles, though :-)

**Comment by moorepants** [comment]
Here is the output from a recent build on the readthedocs system, which is slower than running it locally:

```
computation time summary:
    - ../examples-gallery/plot_parallel_park.py:                             128.56 sec   0.0 MB
    - ../examples-gallery/plot_pendulum_swing_up_variable_duration.py:       127.54 sec   0.0 MB
    - ../examples-gallery/plot_pendulum_swing_up_fixed_duration.py:          124.13 sec   0.0 MB
    - ../examples-gallery/plot_sliding_block.py:                             105.60 sec   0.0 MB
    - ../examples-gallery/plot_ball_rolling_on_spinning_disc.py:              95.81 sec   0.0 MB
    - ../examples-gallery/plot_drone.py:                                      64.82 sec   0.0 MB
    - ../examples-gallery/plot_one_legged_time_trial.py:                      63.60 sec   0.0 MB
    - ../examples-gallery/plot_two_link_pendulum_on_a_cart.py:                49.41 sec   0.0 MB
    - ../examples-gallery/plot_crane_moving_a_load.py:                        33.36 sec   0.0 MB
    - ../examples-gallery/plot_non_contiguous_parameter_identification.py:    12.81 sec   0.0 MB
    - ../examples-gallery/plot_vyasarayani.py:                                12.40 sec   0.0 MB
    - ../examples-gallery/plot_betts2003.py:                                  11.66 sec   0.0 MB
 ```

**Comment by Peter230655** [comment]
If I understand correctly, this is what I tried with the examples that have animations. Should be in the PRs.

**Comment by moorepants** [comment]
Yes, you are. I'm just logging information here on the broader issue. I will get to all your PRs, but it just takes time.

**Comment by Peter230655** [comment]
> Yes, you are. I'm just logging information here on the broader issue. I will get to all your PRs, but it just takes time.

I am in NO rush, I have time! Just wanted to make sure I understood right - which you confirmed!

**Comment by Peter230655** [comment]
> We can add them all as long as we keep the time down and can always delete them later if we every have to replace one.

makes emminent sense!

**Comment by moorepants** [comment]
From c576b31430c782895967fdf5c401208c033072b0:

```
computation time summary:
    - ../examples-gallery/plot_drone.py:                                     251.96 sec   0.0 MB
    - ../examples-gallery/plot_pendulum_swing_up_variable_duration.py:       180.83 sec   0.0 MB
    - ../examples-gallery/plot_ball_rolling_on_spinning_disc.py:             175.27 sec   0.0 MB
    - ../examples-gallery/plot_parallel_park.py:                             165.30 sec   0.0 MB
    - ../examples-gallery/plot_pendulum_swing_up_fixed_duration.py:          162.84 sec   0.0 MB
    - ../examples-gallery/plot_sliding_block.py:                             155.91 sec   0.0 MB
    - ../examples-gallery/plot_mississippi_steamboat.py:                     125.95 sec   0.0 MB
    - ../examples-gallery/plot_one_legged_time_trial.py:                     109.79 sec   0.0 MB
    - ../examples-gallery/plot_car_in_garage.py:                              73.65 sec   0.0 MB
    - ../examples-gallery/plot_crane_moving_a_load.py:                        61.91 sec   0.0 MB
    - ../examples-gallery/plot_two_link_pendulum_on_a_cart.py:                60.65 sec   0.0 MB
    - ../examples-gallery/plot_non_contiguous_parameter_identification.py:    33.34 sec   0.0 MB
    - ../examples-gallery/plot_vyasarayani.py:                                30.82 sec   0.0 MB
    - ../examples-gallery/plot_betts2003.py:                                  25.70 sec   0.0 MB
```

**Comment by moorepants** [comment]
You may have a faster computer than me, but I'm just logging them here manually to keep a check on things. I'm not trying to prompt discussion, just keeping a log.

**Comment by moorepants** [comment]
This is the output from the most recent Github actions CI run:

```
computation time summary:
    - ../examples-gallery/plot_car_around_pylons.py:                         79.64 sec   0.0 MB
    - ../examples-gallery/plot_drone.py:                                     69.66 sec   0.0 MB
    - ../examples-gallery/plot_parallel_park.py:                             65.11 sec   0.0 MB
    - ../examples-gallery/plot_ball_rolling_on_spinning_disc.py:             64.18 sec   0.0 MB
    - ../examples-gallery/plot_sliding_block.py:                             61.86 sec   0.0 MB
    - ../examples-gallery/plot_pendulum_swing_up_variable_duration.py:       61.46 sec   0.0 MB
    - ../examples-gallery/plot_pendulum_swing_up_fixed_duration.py:          59.20 sec   0.0 MB
    - ../examples-gallery/plot_mississippi_steamboat.py:                     44.18 sec   0.0 MB
    - ../examples-gallery/plot_one_legged_time_trial.py:                     41.75 sec   0.0 MB
    - ../examples-gallery/plot_car_in_garage.py:                             30.63 sec   0.0 MB
    - ../examples-gallery/plot_two_link_pendulum_on_a_cart.py:               23.46 sec   0.0 MB
    - ../examples-gallery/plot_crane_moving_a_load.py:                       22.60 sec   0.0 MB
    - ../examples-gallery/plot_betts_10_103_104.py:                          21.71 sec   0.0 MB
    - ../examples-gallery/plot_betts_10_7.py:                                20.33 sec   0.0 MB
    - ../examples-gallery/plot_betts_10_50.py:                               13.38 sec   0.0 MB
    - ../examples-gallery/plot_non_contiguous_parameter_identification.py:   10.93 sec   0.0 MB
    - ../examples-gallery/plot_vyasarayani.py:                               10.25 sec   0.0 MB
    - ../examples-gallery/plot_betts2003.py:                                 10.01 sec   0.0 MB
```

**Comment by moorepants** [comment]
If we are saving solutions, there is no reason to even run `.solve()`.

There are three options a user that downloaded the file may want:

- use our solution without running `.solve()` (same as we want for the docs to run fast)
- use our solution as initial guess and run `.solve()`
- use the initial guess that we actually used to find the solution

If all examples had command line flags like:

```
python plot_example.py --solution=nosolve
python plot_example.py --solution=solutionasguess
python plot_example.py # runs with the original initial guess 
```

or something similar, then I think we can have sphinx gallery pass in the `nosolve` flag. In each example we need something like:

```python
if nosolve:
    solution = np.loadtxt('solution.csv')
elif solutionasguess:
    initial_guess = np.loadtxt('solution.csv')
    solution, info = prob.solve(initial_guess)
else:
    initial_guess = np.zeros(prob.num_free)
    solution, info = prob.solve(initial_guess)
```

We could even make a utility function for that to hide it in the examples.

**Comment by Peter230655** [comment]
If you set ``nosolve`` as default in examples-gallery, this would allow larger simulation to go there.

My feel with many of my simulations was, that the ``animation`` was consuming most of the CPU time.  In most of my examples, I implicitly chose ``solutionasguess`` and still they take time.

**Comment by moorepants** [comment]
Yes, the animations will take time, those have to be generated each time I think. But we can always reduce the number of frames that are generated. 30 fps should be more than sufficient.

**Comment by moorepants** [comment]
I hit my first RTD's CI limit on #361 (ran out of memory). So I think we definitely need to make some changes. For all examples we should at least do this:

```python
if solution_file_exits:
    solution = np.loadtxt(solution_filename)
else:
    initial_guess = ...
    solution, info = problem.solve(initial_guess)
```

This would skip the IPOPT solving entirely on CI if we store a pre-computed solution.

**Comment by Peter230655** [comment]
> I hit my first RTD's CI limit on [#361](https://github.com/csu-hmc/opty/pull/361) (ran out of memory). So I think we definitely need to make some changes. For all examples we should at least do this:
> 
> if solution_file_exits:
>     solution = np.loadtxt(solution_filename)
> else:
>     initial_guess = ...
>     solution, info = problem.solve(initial_guess)
> This would skip the IPOPT solving entirely on CI if we store a pre-computed solution.

I can do this.
solution and info should be in a txt file, not in .npy files as I recall.

**Comment by moorepants** [comment]
I'd prefer to switch to text files, because it is better for git (it is made to track txt not binary files).

**Comment by Peter230655** [comment]
> I'd prefer to switch to text files, because it is better for git (it is made to track txt not binary files).

So I remembered correctly. 
- I guess, solution and info should be stored ? AS I understood np.savetxt(...) it can only save one file?
- This would be one PR for each example?

**Comment by moorepants** [comment]
I think we can skip storing info (see how I did it here #361 )

You can do one PR per example (easiest to review) or more than one in a PR.

**Comment by Peter230655** [comment]
> I think we can skip storing info (see how I did it here [#361](https://github.com/csu-hmc/opty/pull/361) )
> 
> You can do one PR per example (easiest to review) or more than one in a PR.

I prefer one PR per example.
So I'd store info only where it is used explicitly?

**Comment by moorepants** [comment]
If you mean "should I save info to a file on disk"? No we don't need that.

**Comment by moorepants** [comment]
If you are making use of the results of "info" in some way, I would reconsider that. Ideally we save only a single txt file per example (or per solution).

**Comment by Peter230655** [comment]
> If you are making use of the results of "info" in some way, I would reconsider that. Ideally we save only a single txt file per example (or per solution).

Should be possible to get by without the info dict in my examples.

**Comment by moorepants** [comment]
> So should not at least one simple example show how the info dict is used?

Sure that's fine. There are some examples that run super fast and we don't have to store the solutions for them (some of the beginner examples). But for intermediate and advanced we should store the solution and skip solve.

**Comment by moorepants** [comment]
But keep in mind that `info` dict is something from cyipopt (not opty), so we don't technically need to explain it here.

**Comment by Peter230655** [comment]
> But keep in mind that `info` dict is something from cyipopt (not opty), so we don't technically need to explain it here.

Fair enough, but when I started with opty, @tjstienstra  actually gave me the idea, I did not know what was cyipopt or Ipopt, I played around with opty to see what it could do. If opty had a dictionary info, as far as I was concerned it was an opty thing. Maybe other users are similar, though surely much more computer savvy than me.

**Comment by moorepants** [comment]
Yes, I understand and see the value. But that also leads us to copying documentation of other software into our software. The nature of object oriented design causes this big hierarchical tree of software. We don't explain numpy arrays in opty, but opty returns them, for example. You have to draw the line somewhere. Sure we can add a few things that help here and there, but in general we should make pointers to the main source documentation.

**Comment by Peter230655** [comment]
> Yes, I understand and see the value. But that also leads us to copying documentation of other software into our software. The nature of object oriented design causes this big hierarchical tree of software. We don't explain numpy arrays in opty, but opty returns them, for example. You have to draw the line somewhere. Sure we can add a few things that help here and there, but in general we should make pointers to the main source documentation.

I see no need to say more about ``info`` than what is said already in the docstring of solve.
I'd just like to show in one example how it may be used to get infromation. I can look for an example, which runs so fast, no need to store anything.

**Comment by Peter230655** [comment]
When I start doing this, can I at the same time change the docstrings as per your earlier suggestions, or should this better be two different PRs?

**Comment by Peter230655** [comment]
> The simpler PRs are, the easier they are to review and merge quickly.

Makes sense!

**Comment by moorepants** [comment]
Another way we can reduce overall time is to use the new NumPy backend. For many of the simple examples the total time to create the problem + solving it is reduced with NumPy because the problem generation time is significantly shortened.

**Comment by Peter230655** [comment]
Using the kwarg backend = 'numpy' in Problem ?

**Comment by Peter230655** [comment]
I think, most examples are now minimized by 
- providing a solution
- using backend='numpy'
- minimizing fps (frames per second) in the animations

**Comment by Peter230655** [comment]
Could this not be closed?
I guess, we have done whatever we can.

**Comment by Peter230655** [comment]
> @Peter230655 you emailed me and pointed towards #361 where I skipped the solve() (which is what I'm writing that we shouldn't do right now), i.e. I'm being hypocritical! This example and probably the brachistrone still take a long time to solve even if you give the solution as the initial guess. So I think that is why I made that change in #361. It would be best if that is the exception, not the rule.

I NEVER intended to insinuate anything like this!
I wrongly seem to have taken this a the rule to go by.
Now clear to me: solve the problem with a 'perfect' initial guess if it takes a few seconds, skip if it still takes a long time.

**Comment by moorepants** [comment]
I was calling myself hypocritical, rightfully so.

**Comment by Peter230655** [comment]
Just for my understanding:
- does the time to run our simulations figure in here, or is this 'only' the time to build the docs?

**Comment by moorepants** [comment]
This is the total time to build the documentation on ReadTheDocs with 2 cpu cores.

**Comment by Peter230655** [comment]
That is what I meant: we COULD add more simulations if there was a need to show some feature of opty not shown in other examples without running into time-issues.

**Comment by moorepants** [comment]
We are very close to the 15 minute limit, so we may not be able to add any examples until we address that.
